numpy/_core/src/umath/string_ufuncs.cpp

```diff
@@ -345,28 +345,51 @@ string_replace(Buffer<enc> buf1, Buffer<enc> buf2, Buffer<enc> buf3, npy_int64 c
 
     Buffer<enc> end1 = buf1 + len1;
     // Only try to replace if useful.
-    if (len1 >= len2  // Input is big enough to make a match possible.
-        && len2 > 0  // Match string is not empty (so output will be finite).
-        && !(len2 == len3 && buf2.strcmp(buf3) == 0)  // Match and replacement differ.
-        ) {
-        for (npy_int64 time = 0; time < count; time++) {
-            npy_intp pos = findslice_for_replace(buf1, end1 - buf1, buf2, len2);
-            if (pos < 0) {
-                break;
+    if (count > 0        // There's something to replace
+        && len1 >= len2  // Input is big enough to make a match possible.
+        && (len2 > 0 || len3 > 0)  // Match or replacement string is not empty.
+        && !(len2 == len3 && buf2.strcmp(buf3) == 0)) {  // Match and replacement differ.
+
+        if (len2 == 0) {
+            while (count > 0) {  // Interleave
+                buf3.buffer_memcpy(out, len3);
+                out += len3;
+                buf1.buffer_memcpy(out, 1);
+
+                if (--count <= 0) {
```

Comment:
Why not just do `--count` at the end? As is, you'll copy the last character again.

Or, perhaps better, just write it as a `for` loop as for the other branch:
```
for (npy_int64 time = 0; time < count; time ++) {  // Interleave
    buf3.buffer_memcpy(out, len3);
    out += len3;
    buf1.buffer_memcpy(out, 1);
    buf1 += 1;
    out += 1;
}
```
To me, this makes the intent clearer.